### DISCUSSION TITLE
Calling DisposeAsync if Initialization was failed

### DISCUSSION THREAD

**Discussion by strelok372** [xunit/xunit#2313] [Question] [answered]
Hi! I've got unexpected behaviour when I got an exception inside InitializeAsync. 
Maybe I miss something, but I thought that if I will implement IAsyncLifetime with DisposeAsync, then it will be called any time, but in my case it's not. 
Is it correct behaviour?

**Comment by bradwilson** [comment, marked as answer]
Yes, this is the correct behavior.

I modeled `IAsyncLifetime`'s `InitializeAsync` and `DisposeAsync` after the way constructor and `Dispose` works. If the constructor fails, `Dispose` is not called; as such, the same behavior exists with `InitializeAsync` and `DisposeAsync`.

If you do work in `InitializeAsync` that needs to be cleaned up even when it throws, then you should use a `try`/`catch` block to do that cleanup. If your `DisposeAsync` method is written in such a way to clean up a partially initialized object correct, then your `catch` block could be just a simple call to `await DisposeAsync();`.

**Comment by strelok372** [comment]
Oh, you are right, thanks22:04, 12 июня 2021 г., Brad Wilson ***@***.***>:
Yes, this is the correct behavior.
I modeled IAsyncLifetime's InitializeAsync and DisposeAsync after the way constructor and Dispose works. If the constructor fails, Dispose is not called; as such, the same behavior exists with InitializeAsync and DisposeAsync.
If you do work in InitializeAsync that needs to be cleaned up even when it throws, then you should use a try/catch block to do that cleanup. If your DisposeAsync method is written in such a way to clean up a partially initialized object correct, then your catch block could be just a simple call to await DisposeAsync();.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.
--с уважением, Дозоров Владимир Александрович
